### ISSUE TITLE
scrape: Allow to disable end-of-run staleness marker injection for targets 

### ISSUE THREAD

**Issue by thampiotr** [prometheus/prometheus#14049]
## Use case

Suppose a project uses `scrape` package as a library in a cluster where targets are sharded across a number of instances. When a new instance joins the cluster, some portion of targets will typically be handed over to that new instance. The previous owning instance of a target that moved will attempt to inject staleness markers at the end of the scrape loop run (after delay of ~2 intervals). If out-of-order ingestion is enabled, writing staleness markers will succeed and results in occasional gaps in metrics.

[Grafana Alloy](https://github.com/grafana/alloy) faces the above [issue](https://github.com/grafana/alloy/issues/249) in practice. 

## Proposal

Expose a method on scrape manager that allows users to disable end-of-run staleness for user-specified targets. 

In above scenario, users of the `scrape` package will be able to detect that given target has moved to another instance and ensure that there are no unnecessary staleness markers written by the previous owning instance.

More generally, giving users control over when the staleness markers injection should or should not happen, can allow to implement more use cases correctly and with fewer data issues.

**Comment by GiedriusS** [comment]
I guess https://github.com/prometheus/prometheus/issues/13302 is partially related?

**Comment by thampiotr** [comment]
`grafana/prometheus` has [an implementation](https://github.com/grafana/prometheus/pull/35/files) of this which could be upstreamed, unless there some proposals how to do it better

**Comment by tpaschalis** [comment]
Related: https://github.com/prometheus/prometheus/issues/12941

**Comment by beorn7** [comment]
Hello from the bug scrub!

Looking at the older #12941, we stumbled upon this one. We believe an approach like the one proposed here should be tried in Prometheus indeed.

**Comment by fatsheep9146** [comment]
Could I take on this? @beorn7

**Comment by thampiotr** [comment]
@fatsheep9146 thank you for taking interest in this issue, let me know if I can be of any help. 

Currently in Alloy we use a fork [with these changes](https://github.com/grafana/prometheus/pull/34/files) that exposes staleness markers disabling function. This is not the most ergonomic way to do it - I wanted the changes in the fork to be minimal. Creating the right set of `[]*Target` to pass to `DisableEndOfRunStalenessMarkers` is particularly cumbersome. While I would be happy if this approach makes it into Prometheus, it would be great to find a simpler or more ergonomic way to do it.

There is one alternative that I was thinking about. Maybe @beorn7 can chime in if it's feasible? For the use case where you have multiple clients remote_writing to storage backend, you would typically want to have the out-of-order window enabled too, because target handovers and write race between instances. If that's the case, perhaps we could leave the remote_write clients' staleness handling as-is, but discard the unwanted staleness markers during out-of-order window processing in the backend? Perhaps the out-of-order samples/histograms can be allowed, but out-of-order staleness markers should be rejected? 

Furthermore, I can see [some mentions in scrape.go](https://github.com/prometheus/prometheus/blob/main/scrape/scrape.go#L1533) that appear to rely on OOO staleness markers being rejected, but AFAICT it's not going to happen if you enable OOO window - not sure if this logic works correctly right now when OOO is enabled. 

The above approach would make staleness handling a bit simpler in a cluster of remote_write clients and could help with other use cases, like the one I've mentioned in [this Alloy proposal](https://github.com/grafana/alloy/issues/2547). But maybe I'm missing some important context here, so would be great to hear from both of you what you think.

**Comment by bboreham** [comment]
> Furthermore, I can see [some mentions in scrape.go](https://github.com/prometheus/prometheus/blob/main/scrape/scrape.go#L1533) that appear to rely on OOO staleness markers being rejected, but AFAICT it's not going to happen if you enable OOO window - not sure if this logic works correctly right now when OOO is enabled.

This was addressed by https://github.com/prometheus/prometheus/pull/14710

**Comment by beorn7** [comment]
Sorry for not being more knowledgeable about the remote-write situation. Does #14710 remove the reason why we have this feature request?

If this feature request is still needed, @fatsheep9146 is of course free to go ahead, and I would love if @thampiotr could help with the review to make sure the implementation works as intended.

**Comment by fatsheep9146** [comment]
@thampiotr Thanks!

**Comment by thampiotr** [comment]
> Does https://github.com/prometheus/prometheus/pull/14710 remove the reason why we have this feature request?

No, but it does remove the concern I've raised here: "Furthermore, I can see [some mentions in scrape.go](https://github.com/prometheus/prometheus/blob/main/scrape/scrape.go#L1533) that appear to rely on OOO staleness markers being rejected, but AFAICT it's not going to happen if you enable OOO window"

I still would love to hear what more experience Prometheus contributors think about the general idea of "leave the remote_write clients' staleness handling as-is, but discard the unwanted staleness markers during out-of-order window processing in the backends" as described in [this comment](https://github.com/prometheus/prometheus/issues/14049#issuecomment-2696995889).

**Comment by krajorama** [comment]
> There is one alternative that I was thinking about. Maybe @beorn7 can chime in if it's feasible? For the use case where you have multiple clients remote_writing to storage backend, you would typically want to have the out-of-order window enabled too, because target handovers and write race between instances. If that's the case, perhaps we could leave the remote_write clients' staleness handling as-is, but discard the unwanted staleness markers during out-of-order window processing in the backend? Perhaps the out-of-order samples/histograms can be allowed, but out-of-order staleness markers should be rejected?

I've rewatched @brian-brazil 's [video on staleness](https://www.youtube.com/watch?v=GcTzd2CLH7I) to get into this topic again. The relevant parts are the "Target goes away" discussion. Without clustering Prometheus, the solution for target moving between scrape configs is to rely on TSDB not having out of order ingestion . Basically assume that if the target moved (as opposed to going away), we'll not be able to write the staleness marker because there's a newer sample. Before OOO it was trivial, after OOO it is also works with the fix in #14710. 

Let's look at the use case of @thampiotr where there's an original Prometheus instance and a new Prometheus instance and the target moves from the original to the new one. I'll shorten scrape interval to "si".

Since we're dealing with independent processes and schedulers, anything can happen:

1. Original has stopped scraping the target. But the new never comes up.
   In this case the original will eventually write a staleness marker to remote , which is fine.
2. Original has stopped scraping the target, sends a staleness marker for timestamp (t+si). The new instance starts scraping the target and sends a sample to remote.
   1. New instance is fast, send is fast, writes a sample for (t+y) < (t+si), and it arrives before the staleness marker. This is not OOO. (This is somewhat unlikely as the new scrape loop needs to start up in the new instance.)
   2. New instance is fast, send is slow, write a sample for (t+y) < (t+si), but arrives after the staleness marker. This is OOO for the new instance. (This is unlikely as the staleness marker is actually sent at t+2*si*1.1.)
   3. New instance is slow, write fast, write sample for (t+y) > (t+si). The staleness marker arrives after. This is OOO for the staleness marker. (Common case.)
   4. New instance is slow, writes slow., write sample for (t+y) > (t+si). The staleness marker arrives before the sample. Not OOO. (Seems unlikely as the staleness marker is actually sent at t+2*si*1.1.)

From the remote write receiver point of view some of these cases where we don't have OOO seem hard to distinguish from real cases. In case of OOO: the OOO window can be 24h so we might in fact get legit staleness markers from hours ago. If we knew the scrape interval in the server, we might have some idea what to reject. 

I haven't really come to any conclusions yet and I'm not even sure about the probabilities in the cases, but this in itself suggests that simply turning off staleness markers and relying on the sender being HA could be a viable solution.
